### DISCUSSION TITLE
OpenSP on Apple Silicon issue

### DISCUSSION THREAD

**Discussion by wrobelda** [microsoft/vcpkg#20347] [General]
I am working on OpenSP port and it is failing to compile statically on macOS on Apple Silicon with:

```
autoreconf --verbose --install --force
./configure --disable-debug --disable-dependency-tracking --enable-http --disable-doc-build --disable-shared --enable-static
make
```

```
/bin/sh ../libtool  --tag=CXX   --mode=link g++  -g -O2   -o onsgmls nsgmls.o SgmlsEventHandler.o RastEventHandler.o StringSet.o nsgmls_inst.o ../lib/libosp.la ../intl/libintl.a -liconv -lpthread
libtool: link: g++ -g -O2 -o onsgmls nsgmls.o SgmlsEventHandler.o RastEventHandler.o StringSet.o nsgmls_inst.o -Wl,-bind_at_load  ../lib/.libs/libosp.a ../intl/libintl.a -liconv -lpthread
ld: warning: ignoring file ../lib/.libs/libosp.a, building for _macOS-arm64_ but attempting to link with file built for _macOS-arm64_
Undefined symbols for architecture arm64:
```

Notice the puzzling error message: **building for macOS-arm64 but attempting to link with file built for macOS-arm64**

I inspected the `libosp.a` and it is indeed macOS-arm64, which would suggest some issue with libtool. I then attempted to build the same with brew, having edited the recipe and added  `--disable-shared --enable-static` and somehow it builds just fine. Brew has a dependency on its own `libtools`, so I exported it to path per their instructions (`export PATH="$(brew --prefix)/opt/libtool/libexec/gnubin:$PATH"`) yet it still fails.

Anyone has any idea how to proceed here?

**Comment by autoantwort** [comment]
Maybe brew runs as arm64 and vcpkg runs as x64 and it makes a difference. Or the other way around. You can check it with `lipo -archs ...`

**Reply by wrobelda** [reply]
At this point it has nothing to do with vcpkg – I am compiling it barebone, with those 3 commands only (autoreconf, configure, make)

**Reply by autoantwort** [reply]
Ah ok :)

**Reply by autoantwort** [reply]
Have not read it completely but I think this could work https://github.com/numpy/numpy/issues/18523

**Reply by autoantwort** [reply]
Or rather https://github.com/bitcoin/bitcoin/issues/20825#issuecomment-753444519

**Reply by wrobelda** [reply]
Thanks, but saw both already, and am using Apple's `ld` and `ranlib`:

```
 cromo@Daves-Macbook-Air   ranlib -V                                                
Apple Inc. version cctools-980
 cromo@Daves-Macbook-Air   ld -v                                              
@(#)PROGRAM:ld  PROJECT:ld64-650.9
BUILD 00:19:34 Mar 17 2021
configured to support archs: armv6 armv7 armv7s arm64 arm64e arm64_32 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em
LTO support using: LLVM version 12.0.5, (clang-1205.0.22.9) (static support for 27, runtime is 27)
TAPI support using: Apple TAPI version 12.0.5 (tapi-1205.0.7.1)

**Reply by wrobelda** [reply]
The last resort is for someone with M1 as well to test it on their end.

**Reply by autoantwort** [reply]
Yeah I can do it when you send me the commands I have to execute

**Reply by wrobelda** [reply]
Really just these three:
```
autoreconf --verbose --install --force
./configure --disable-debug --disable-dependency-tracking --enable-http --disable-doc-build --disable-shared --enable-static
make
```

**Reply by autoantwort** [reply]
And a git clone?

**Reply by wrobelda** [reply]
A tarball: https://sourceforge.net/projects/openjade/files/opensp/1.5.2/OpenSP-1.5.2.tar.gz/download

**Reply by autoantwort** [reply]
Yeah same error

**Reply by wrobelda** [reply]
Yet somehow brew can compile it statically just fine, so I am assuming it must be its own libtool or other magic they do in their compiler sandbox. 

EDIT: although, as I mentioned, I tried with their libtool to no avail.

**Reply by autoantwort** [reply]
They don't pass `--disable-shared --enable-static` in their script. If I remove that it also works for me

**Reply by wrobelda** [reply]
> They don't pass `--disable-shared --enable-static` in their script. If I remove that it also works for me

_"I then attempted to build the same with brew, having edited the recipe and added --disable-shared --enable-static and somehow it builds just fine."_

Try `brew edit open-sp` and adding those options by hand. Then `brew reinstall --build-from-source open-sp`

**Reply by autoantwort** [reply]
Oh sorry didn't read that. Yeah it works with brew

**Reply by wrobelda** [reply]
FYI https://github.com/Homebrew/discussions/discussions/2207

**Reply by wrobelda** [reply]
OK, finally figured it out. 

1. Brew sets the `HOMEBREW_LIBRARY_PATHS` env variable, which contains `/opt/homebrew/lib` and is added to LDFLAGS by their clang wrapper. 
2. That, in turns, links against their `libintl`.
3. If pointed to brew's `libintl` by adding `-with-libintl-prefix=/opt/homebrew/` to `./configure`, the vanilla source compiles OK. 
4. Their lib, however, is shared. So to the reason vcpkg fails against its own `libintl` is because it is static and needs extra `-framework CoreFoundation` LDFLAG, otherwise it throws bunch of `Undefined symbols for architecture arm64:  "_CFArrayGetCount"` errors when linking.
5. This was all in the initial vcpkg's log file 🤦🏻‍♂️
